Optimize PostgreSQL Settings for 4x12GB Applaince

rvalente opened the pull request:
These changes greatly improve the performance of PostgreSQL, specifically maintenance tasks such as `VACUUM` and `ANALYZE`.

BZ: ~~https://bugzilla.redhat.com/show_bug.cgi?id=1547161~~
https://bugzilla.redhat.com/show_bug.cgi?id=1561112

rvalente committed:
Optimize PostgreSQL Settings for 4x12GB Applaince

tune wal_buffer, min_wal_size, and max_wal_size

carbonin reviewed (dismissed):
Changes from the defaults shipped by the postgres version should be made in our included file [here](https://github.com/ManageIQ/manageiq-appliance/blob/master/COPY/etc/manageiq/postgresql.conf.d/01_miq_overrides.conf)

psuriset commented:
Thanks for sharing this fix. I would change shared_buffer, based on the scale user runs.  I feel, its challenging to dedicate 3G shared_buffer for  an appliance with default memory.

rvalente commented:
@carbonin updated PR to put changes in the correct location.

@psuriset We run our setup with only the `database operations` role on the PostgreSQL server. I realize that may not be the norm, I have updated the tuning settings to assume we only have 8GB of memory for the PostgreSQL server. This tightens up the `work_mem`, `shared_buffers`, and `maintenance_work_mem`.

I don't know what the standard usage looks like in terms of connections, but we have reliably stayed under 500 connections, even under our largest deployments. Dropping `max_connections` from 1000 to 500 allows for more memory to be used per connection. Typically I don't like to exceed that many concurrent connections on a single PostgreSQL box, I use [pgpool](http://www.pgpool.net/mediawiki/index.php/Main_Page) if more connections are needed.

rvalente commented:
Following on to what @psuriset is saying: it would be fantastic to leverage Ansible to tune the `postgresql.conf` *after* the appliance is deployed based on the hardware and roles that the server is running.

That would allow tweaks even after the appliance is deployed, to ensure as an environment grows in managed object count, the DB is able to keep up and stay healthy.

Minimum viable implementation, would be shipping multiple configs: (dedicated, small, large) since a single `postgresql.conf` can't be one-size fits all.

carbonin commented:
> We run our setup with only the database operations role on the PostgreSQL server. I realize that may not be the norm, I have updated the tuning settings to assume we only have 8GB of memory for the PostgreSQL server.

We try to alter these settings if we see that a server is configured as a "database only" appliance [here](https://github.com/ManageIQ/manageiq-appliance_console/blob/master/lib/manageiq/appliance_console/internal_database_configuration.rb#L169), maybe the values we're using there could use an update as well.

> I don't know what the standard usage looks like in terms of connections, but we have reliably stayed under 500 connections, even under our largest deployments.

What do your "largest deployments" tend to look like? How many appliances/workers per region?

> Typically I don't like to exceed that many concurrent connections on a single PostgreSQL box, I use pgpool if more connections are needed.

I tend to agree. We've investigated deploying a connection pool, but managing that programatically is quite a task. If users want to implement their own connection pools to reduce this value, I don't think that would be a problem.

> it would be fantastic to leverage Ansible to tune the postgresql.conf after the appliance is deployed based on the hardware and roles that the server is running.

Maybe we can do something like this in the area I mentioned above. If we see that the server is going to be only running the database, we can tune it using some standard calculations. Are the methods you used to come to these values easily written up in code?

> That would allow tweaks even after the appliance is deployed, to ensure as an environment grows in managed object count, the DB is able to keep up and stay healthy.

> Minimum viable implementation, would be shipping multiple configs: (dedicated, small, large) since a single postgresql.conf can't be one-size fits all.

This is exactly the use case we envisioned when adding the `conf.d` directory. This way we can keep the base file up to date with the defaults shipped with the PG version we're using, we can keep this file up to date with what we think is the 80% case or so, then the user can put their environment-specific tuning in later files in this directory or use postgresql's [`ALTER SYSTEM`](https://www.postgresql.org/docs/9.5/static/sql-altersystem.html) command to change things to fit their particular usage pattern. 

I'm not sure why you would need Ansible for this, but if that's your tool of choice, I'm sure it would do the job.

-------------------------------------------

As for this patch, can you provide the methods you used for coming to these values and a description of the environment that you used them in? Some context will really help in evaluating if these are sensible defaults for all users.

rvalente commented:
>> We run our setup with only the database operations role on the PostgreSQL server. I realize that may not be the norm, I have updated the tuning settings to assume we only have 8GB of memory for the PostgreSQL server.

> We try to alter these settings if we see that a server is configured as a "database only" appliance here, maybe the values we're using there could use an update as well.

I will update that as well, and try and bake in some logic for dedicated/shared combined with actual memory capacity on the appliance.

>> I don't know what the standard usage looks like in terms of connections, but we have reliably stayed under 500 connections, even under our largest deployments.

> What do your "largest deployments" tend to look like? How many appliances/workers per region?

Global Roll-up Region
Site 1 (2 Providers, 4 Clusters, 34 Hosts, 80 Datastores, 9464 VMs, 227 Templates) 9809 Objects
Site 2 (2 Providers, 3 Clusters, 50 Hosts, 119 Datastores, 7600 VMs, 957 Templates) 8729 Objects
Site 3 (1 Provider, 10 Clusters, 33 Hosts, 49 Datastores, 4303 VMs, 773 Templates) 5168 Objects

![cfme_roles](https://user-images.githubusercontent.com/5966/36547806-b2a9a806-17bc-11e8-85f7-5d1ad95288a8.png)

I created "groups" that are associated with appliances in our environment. Each group has specific roles associated with them to standardize the deployment and keep everything consistent across our global footprint. Using `configure_server_settings.rb` and Ansible we manage our ~110 appliances across 4 Regions today. Long term architecture for this deployment adds an additional 10 Regions. So as you could imagine, making these changes manually is not feasible.

```yaml
- hosts: cfme-ui

vars:
  cfme_roles:
    - notifier
    - scheduler
    - user_interface
    - web_services
    - websocket
    - ems_operations
    - reporting

tasks:
  - name: Assign Roles from Vars
    command: ./tools/configure_server_settings.rb -s {{ cfme_server_id }} -p server/role -v {{ cfme_roles|join(',') }}
    args:
      chdir: "{{ cfme_approot }}"
```

>> Typically I don't like to exceed that many concurrent connections on a single PostgreSQL box, I use pgpool if more connections are needed.

> I tend to agree. We've investigated deploying a connection pool, but managing that programmatically is quite a task. If users want to implement their own connection pools to reduce this value, I don't think that would be a problem.

>> it would be fantastic to leverage Ansible to tune the `postgresql.conf` after the appliance is deployed based on the hardware and roles that the server is running.

> Maybe we can do something like this in the area I mentioned above. If we see that the server is going to be only running the database, we can tune it using some standard calculations. Are the methods you used to come to these values easily written up in code?

[Here](http://pgtune.leopard.in.ua/) is a great resource based on the original [pgtune code](https://github.com/gregs1104/pgtune).

>> That would allow tweaks even after the appliance is deployed, to ensure as an environment grows in managed object count, the DB is able to keep up and stay healthy.

>> Minimum viable implementation, would be shipping multiple configs: (dedicated, small, large) since a single `postgresql.conf` can't be one-size fits all.

> This is exactly the use case we envisioned when adding the conf.d directory. This way we can keep the base file up to date with the defaults shipped with the PG version we're using, we can keep this file up to date with what we think is the 80% case or so, then the user can put their environment-specific tuning in later files in this directory or use postgresql's `ALTER SYSTEM` command to change things to fit their particular usage pattern.

My only concern with this is it potentially obfuscates the configuration method if changes are both made to the appliance. So both options are sufficient, but I personally would select only a single option. If `ALTER SYSTEM` is used then the changes made would be located in `postgresql.auto.conf`.

> I'm not sure why you would need Ansible for this, but if that's your tool of choice, I'm sure it would do the job.

Agreed, Ansible is not required here by any stretch but it is one potential solution. If `ALTER SYSTEM` is used then no cfg. mgmt solution would be required.

> As for this patch, can you provide the methods you used for coming to these values and a description of the environment that you used them in? Some context will really help in evaluating if these are sensible defaults for all users.

My initial debugging started by debugging the ~200 checkpoint errors in the logs. That was the initial tuning step of increasing `max_wal_size` and `wal_buffers`. While this reduced the occurrence of the checkpoint errors in the logs, it didn't fix the 100% swap issue and the inability for PostgreSQL to keep up with the metrics purging and `AUTOVACUUM`.

After additionally digging using the PostgreSQL tuning wiki and the pgtune tool linked above, I made the adjustments to more closely align PostgreSQL with the hardware that it was running on. 

Changes and reasoning listed below:

----

`max_connections = 500` : even at our large deployments

`shared_buffers = 2GB` : "reasonable starting value for shared_buffers is 1/4 of the memory in your system"[1](https://wiki.postgresql.org/wiki/Tuning_Your_PostgreSQL_Server) - reduced by 1GB for `evmserverd` headroom.

`work_mem = 4194kB` : Larger `work_mem` will allow for complex sorts to occur in-memory resulting in better performance.

`maintenance_work_mem = 512MB` : Required a large bump to allow for `VACUUM` and `CREATE INDEX` performance increase.

`min_wal_size = 1GB` : Checkpoint errors in logs fix - tune for checkpoint optimization

`max_wal_size = 2GB` : Checkpoint errors in logs fix - tune for checkpoint optimization

`checkpoint_completion_target = 0.7` : this can be increased back up to 0.9, however changes to `checkpoint_segments` will be [recommended](http://www.postgresql.org/docs/current/static/runtime-config-wal.html#RUNTIME-CONFIG-WAL-CHECKPOINTS)

`random_page_cost = 4` : I dropped this to 1.1 in our environment because we're stored on SSD. This is a cost value of disk seek to help the query optimizer.

`effective_cache_size = 6GB` : "Setting effective_cache_size to 1/2 of total memory would be a normal conservative setting, and 3/4 of memory is a more aggressive but still reasonable amount."[1](https://wiki.postgresql.org/wiki/Tuning_Your_PostgreSQL_Server)

`default_statistics_target = 100` : this is the new default in PG 8.4 and later, and can be removed from the PR for improved readability.

[1] [https://wiki.postgresql.org/wiki/Tuning_Your_PostgreSQL_Server](https://wiki.postgresql.org/wiki/Tuning_Your_PostgreSQL_Server)

carbonin commented on COPY/etc/manageiq/postgresql.conf.d/01_miq_overrides.conf:
```diff
@@ -11,7 +11,7 @@
 
 listen_addresses = '*'
-max_connections = 1000
+max_connections = 500
```
While I agree that this shouldn't be this high. I don't think I'm comfortable lowering it.

This is more of a compatibility issue for me.
If we lower this and a user happens to be over 500, that's a pretty serious breaking change. I would rather leave it up to each user to decide where this should be now that it is this high.

Also, what downside is there to having a higher upper limit? For example, if we only ever establish 200 connections, does having this number at 1000 incur any overhead?

rvalente replied:
It does, max_connections has a direct impact on `work_mem` tuning. This is where some dynamic tuning based on worker and process could would be beneficial. I will set this back to 1000

> This is very important to some of the below parameters (particularly work_mem) because there are some memory resources that are or can be allocated on a per-client basis, so the maximum number of clients suggests the maximum possible memory use. Generally, PostgreSQL on good hardware can support a few hundred connections.

carbonin commented on COPY/etc/manageiq/postgresql.conf.d/01_miq_overrides.conf:
```diff
@@ -26,6 +26,9 @@ tcp_keepalives_interval = 75
 #------------------------------------------------------------------------------
 
+shared_buffers = 2GB
+work_mem = 4194kB
```
Isn't the default [4MB](https://www.postgresql.org/docs/9.5/static/runtime-config-resource.html#GUC-WORK-MEM)? Is the extra 200KB really making a difference here?

rvalente replied:
On a 6GB appliance, with 1k connections, `work_mem` should only be `1572kB`.
Dropping connections to 500 would enable `work_mem = 3145kB`.

carbonin commented on COPY/etc/manageiq/postgresql.conf.d/01_miq_overrides.conf:
```diff
@@ -26,6 +26,9 @@ tcp_keepalives_interval = 75
 # RESOURCE USAGE (except WAL)
 #------------------------------------------------------------------------------
 
+shared_buffers = 2GB
```
Can we remove this in favor of https://github.com/ManageIQ/manageiq-appliance_console/pull/33 ?

rvalente replied:
We can, but it isn't immediately obvious that this is changed via `appliance_console` on deployment. Removing.

carbonin commented on COPY/etc/manageiq/postgresql.conf.d/01_miq_overrides.conf:
```diff
@@ -26,6 +26,9 @@ tcp_keepalives_interval = 75
 
+shared_buffers = 2GB
+work_mem = 4194kB
+maintenance_work_mem = 512MB
```
This is per autovacuum worker. There are 3 of those by default. So if we want this value up this high we should set [autovacuum_work_mem](https://www.postgresql.org/docs/9.5/static/runtime-config-resource.html#GUC-AUTOVACUUM-WORK-MEM) to something a bit smaller I think. Maybe 128MB (twice the default)?

rvalente replied:
Good idea just to be safe. However, this is where most of the benefit of this tuning was realized. These tuning options combined with the truncate PR will be a great combination of improvements. I'll update.

carbonin commented on COPY/etc/manageiq/postgresql.conf.d/01_miq_overrides.conf:
```diff
@@ -36,7 +39,9 @@ max_worker_processes = 10
-checkpoint_completion_target = 0.9
+min_wal_size = 1GB
+max_wal_size = 2GB
+checkpoint_completion_target = 0.7
```
> checkpoint_completion_target = 0.7 : this can be increased back up to 0.9, however changes to checkpoint_segments will be recommended

I think I'd rather keep this at 0.9 and `checkpoint_segments` was removed in 9.5

rvalente replied:
👍

carbonin commented on COPY/etc/manageiq/postgresql.conf.d/01_miq_overrides.conf:
```diff
@@ -51,6 +56,9 @@ hot_standby = on
 #------------------------------------------------------------------------------
 
+random_page_cost = 4
+effective_cache_size = 6GB
```
ManageIQ appliances only ship with 6GB of RAM, so this would be too high I think.

Also the default is 4GB so we're already over what your recommendation (1/2 of total memory) would be.

rvalente replied:
Dropped to the recommended 4608MB.

carbonin commented on COPY/etc/manageiq/postgresql.conf.d/01_miq_overrides.conf:
```diff
@@ -51,6 +56,9 @@ hot_standby = on
 
+random_page_cost = 4
+effective_cache_size = 6GB
+default_statistics_target = 100
```
This is the default so we can remove this.

rvalente replied:
👍

carbonin reviewed (commented):
Thanks for the thorough write up @rvalente ! It was definitely helpful in understanding where you're coming from.

I agree with most of the changes, but I think we'll have to tone some down a bit as the ManageIQ appliance ships with only 6GB of memory.

Added individual comments inline.

rvalente commented:
@carbonin the more I think about this the more I agree with you regarding how to make these changes. Making these 1 size fits most changes in here is fairly useless. That said, deploying an appliance implies that it will support the max configurations.

So that being said, I want to drop these changes down to just the min/max wal_buffers to get this merged in and then focus on a more host specific tune done via appliance_console.

Thoughts?

carbonin commented:
I think that makes a lot of sense @rvalente 

Fix the bug you encountered with this PR, then we can either encourage people to tune their environments after deploy or do something more clever in the console where we might have more information about the system we are running on.

carbonin commented:
@rvalente could you squash the previous commits?

rvalente commented:
@carbonin yep! done and done

carbonin reviewed (approved):
Looks good, thanks for this @rvalente the change is good, and the discussion was worth it as well :smile:

JPrause commented:
@miq-bot add_label blocker

JPrause commented:
@rvalente can you add the gaprindashvili/yes and fine/yes labels   copying @carbonin

rvalente commented:
@JPrause Doesn't look like I have the permission to add labels to this PR.

rvalente commented:
For anyone looking at this PR in the future, we have removed the following tune from our PG deployment:, we ran into replication issues with the global reporting region.

```
random_page_cost = 4 : I dropped this to 1.1 in our environment because we're stored on SSD. This is a cost value of disk seek to help the query optimizer.
```

JPrause commented:
@rvalente I can add it,..but I always want to make sure PRs can be backported with the developer. Sounds like it can be, so I'll add the labels.

JPrause commented:
@miq-bot add_label gaprindashvili/yes, fine/yes

rvalente commented:
Sounds good, thanks!

JPrause commented:
@simaishi Related BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1547161

simaishi commented:
This PR actually goes with https://bugzilla.redhat.com/show_bug.cgi?id=1561112 (Thanks @carbonin for explaining the issues and related PRs), and it's not a blocker.

simaishi commented:
@carbonin @dmetzger57 I assume we don't want to backport this to Gaprindashvili/Fine yet (until we know more about what to tune)?

dmetzger57 commented:
Correct

Sent from my iPhone

On Apr 16, 2018, at 5:59 PM, Satoe Imaishi <notifications@github.com> wrote:

@carbonin <https://github.com/carbonin> @dmetzger57
<https://github.com/dmetzger57> I assume we don't want to backport this to
Gaprindashvili/Fine yet (until we know more about what to tune)?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/ManageIQ/manageiq-appliance/pull/181#issuecomment-381763749>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ALtde36fz-8o4ltdhK5DM7KeXcTwKvH7ks5tpRRSgaJpZM4SOX-s>
.
